File: tools/generate_checkpoint_fixtures.py
"""Write a partial checkpoint with the Python client (for the R client's cross-language test).

    .venv/bin/python tools/generate_checkpoint_fixtures.py
"""

import json
import pathlib
import shutil
import sys

REPO = pathlib.Path(__file__).resolve().parents[1]
sys.path.insert(0, str(REPO / "python" / "src"))
sys.path.insert(0, str(REPO / "python"))

from llikert import Scorer, ScoringTask  # noqa: E402
from tests.client.mock_service import MockService  # noqa: E402

OUT = REPO / "tests" / "fixtures" / "checkpoints" / "python-partial"
mock = MockService()
scorer = Scorer.connect("http://mock", _client=mock.client())
task = ScoringTask(
    name="Communicative function",
    instructions="Classify the text's primary communicative function.",
    categories=["description", "question", "request"],
    responses=["A", "B", "C"],
)
prepared = scorer.prepare(task)
dataset = json.loads((REPO / "tests/fixtures/runs/dataset.json").read_text())

original = mock.score
calls = {"n": 0}


def crash_after_two(request):
    calls["n"] += 1
    if calls["n"] > 2:
        raise RuntimeError("simulated crash")
    return original(json.loads(request.content))


for _ in range(3):
    mock.inject["/v1/score"].append(crash_after_two)
shutil.rmtree(OUT, ignore_errors=True)
try:
    scorer.score(dataset["texts"], dataset["ids"], task=prepared, chunk_size=5, checkpoint=OUT, progress=False)
except RuntimeError:
    pass
print("wrote", OUT, sorted(p.name for p in (OUT / "chunks").iterdir()))

File: tools/generate_checkpoint_fixtures.R
# Write a partial checkpoint with the R client (for the Python client's cross-language test).
#
#   Rscript tools/generate_checkpoint_fixtures.R      (from the repository root)

devtools::load_all("r/llikert", quiet = TRUE)
source("r/llikert/tests/testthat/helper-mock.R")
fixture_path <- function(...) file.path("tests/fixtures", ...)

out <- "tests/fixtures/checkpoints/r-partial"
unlink(out, recursive = TRUE)
mock <- local_mock()
engine <- scorer_connect("http://mock")
prepared <- prepare_task(quickstart_nominal(), engine)
d <- dataset()
count <- 0L
crash <- function(req, body) {
  count <<- count + 1L
  if (count > 2L) stop("simulated crash")
  mock$score(body)
}
for (i in 1:3) mock$push("/v1/score", crash)
try(score_items(d$texts, d$ids, task = prepared, engine = engine, chunk_size = 5L, checkpoint = out, progress = FALSE), silent = TRUE)
cat("wrote", out, list.files(file.path(out, "chunks")), "\n")
